### ISSUE TITLE
lists in input source field can cause trouble

### ISSUE THREAD

**Issue by KerstenBreuer** [rabix/composer#339]
I experience very uninformative errors when running my rabix-compsed workflow with toil.
As I reported in the corresponding issue (https://github.com/BD2KGenomics/toil/issues/2022), the cause is that rabix composer specifies the input sources as lists:

```
steps:
  validate_data:
    in:
      feat_in:
        source: 
          - feat_in
      label_in:
        source: 
          - label_in
      metadata_in:
        source: 
          - metadata_in
    out:
      - validated_feat
      - validated_label
      - validated_metadata
    run: ../tools/01_validate_data.cwl
```
However, toil only likes this if you use the `linkMerge` method, which of cause makes sense. Even though cwltool seems to have no problem with this, I think it would make sense to change the behavior to this:
```
steps:
  validate_data:
    in:
      feat_in:
        source: feat_in
      label_in:
        source: label_in
      metadata_in:
        source: metadata_in
    out:
      - validated_feat
      - validated_label
      - validated_metadata
    run: ../tools/01_validate_data.cwl
```
(The same also applies to the `outputSource` fields.)

Or is there another case in which listing of inputs might be useful which I am missing right now?

**Comment by mayacoda** [comment]
The one thing I'm not really clear on in the CWL specification is whether `linkMerge` can be defined if there is only one inbound data link (i.e. that `source` is a single item, not an array). @mr-c or @tetron, could you please help out and clarify? Thanks!

This issue arose because, for convenience, I always had the [rabix/cwl-ts](https://github.com/rabix/cwl-ts) model serialize to the "more inclusive" data type. However, it's evident that in some cases data types change behavior, so a fair warning to my successors :)

As for this bug, if `linkMerge` still works when `source`/`outputSource` is a single item, the safest and easiest solution is to always serialize these fields as single times instead of arrays of one item.

**Comment by tetron** [comment]
We've been discussing this over here on the spec repo.

https://github.com/common-workflow-language/common-workflow-language/issues/675

I only just now saw this Composer ticket (I should have checked earlier).  I hope someone from the Composer team can join the next CWL call so we can sort this out?

**Comment by ivanbatic** [comment]
We added a linkMerge selection dropdown to the input definition panel in the workflow editor, so users can explicitly see the setting.

**Comment by tetron** [comment]
@ivanbatic could you read/respond on the ticket I linked above?

**Comment by kaushik-work** [comment]
@marijanlekic did we merge this fix from @bogdang989 into the current round? It is the source of some annoyance. Just putting this here in case I forget.

**Comment by marijanlekic** [comment]
@kaushik-work 

No, pull request fixes a small issue regarding arguments but does not fix the issue related to source problem. I'll look at the code to see where is the root of the problem but we will see if its going to be included in upcoming bug fixes release.

**Comment by marijanlekic** [comment]
@kaushik-work After further analysis with @bogdang989 it looks like that when using linkMerge with value "merge_nested" then its working, thats why its added. I suppose we can close this issue.

**Comment by mr-c** [comment]
Is the solution from @mayacoda implemented, @marijanlekic, or is this something else?

> As for this bug, if `linkMerge` still works when source/outputSource is a single item, the safest and easiest solution is to always serialize these fields as single times instead of arrays of one item.

**Comment by marijanlekic** [comment]
@mr-c 

If "Nested" is selected in a dropdown then ["input"] will be "input" when serialized. When "Flattened" is selected then ["input"] will be ["input"].  This is implemented a while ago. Is this fixing the issue?

**Comment by mr-c** [comment]
@marijanlekic What is the default value of the dropdown?

**Comment by marijanlekic** [comment]
@mr-c 

"Nested"

**Comment by mr-c** [comment]
@marijanlekic To confirm, the drop down has two states and one of them must be chosen: "nested" and "flattened"

`linkMerge` has three possible states: not present, "nested" and "flattened"

If a single output is connected to an input port of another step it should not be wrapped in an array by default, nor should `linkMerge` be present, unless the user has proactively chosen to add `linkMerge`

To summarize: with single inputs, `linkMerge: nested` does not produce the same behaviour in CWL as no `linkMerge` at all.

**Comment by kaushik-work** [comment]
Thanks to a mathematical definition from @mr-c over gitter we've been able to create a table. Now gitter isn't specifically designed to convey this kind of information, so here is the logic table as I understood it for review.


|  |  | Scalar Sink | Scalar Sink | Scalar Sink | Array Sink | Array Sink | Array Sink |
| -- | -- | -- | -- | -- | -- | -- | -- | 
|  |   | LinkMerge None | LinkMerge Nested | LinkMerge Flattened | LinkMerge None | LinkMerge Nested | LinkMerge Flattened |
| Single Source | Scalar Source | Default | ILLEGAL | ILLEGAL | ILLEGAL | Default | Valid |
| Single Source | Array Source | Default | ILLEGAL | ILLEGAL | ILLEGAL | Default | Valid |
| Multiple Sources | Scalar Source | requires valueFrom or scatter | requires valueFrom or scatter | requires valueFrom or scatter | ILLEGAL | Default | Valid |
| Multiple Sources | Array Source | requires valueFrom or scatter | requires valueFrom or scatter | requires valueFrom or scatter | ILLEGAL | Default | Valid |

**However, I'm not clear if the values are indeed independent of the source type (Scalar vs Array for source) as I've written**

**Comment by kaushik-work** [comment]
Ok, the full LinkMerge matter seems wider than this issue. This issue we will resolve in the following manner
- add a "None" option to the LinkMerge methods list (d6e59778489b3f9bd0d24ddbb55a382560446b8b)
- add `MultipleInputFeatureRequirement` when relevant
- When a scalar (non-array) source is connected to a scalar (non-array) sink do not wrap that in an array

The wider matter will be followed in the linked tickets

**Comment by kaushik-work** [comment]
Quick quality check using commit d6e59778489b3f9bd0d2 :

* "None" option has been added to LinkMerge dropdown
* This is the default
* single scalar source connected to a port will result in a simply (non array) entry in CWL code
* Bi-directionality works correctly - changing code will change drop down value
* User can intentionally add LinkMerge to single inputs. This is a user error. UI will not stop this intentionally incorrect behavior
* If user enters incorrect CWL in `code` tab  (e.g. mis-spelling `merge_nested`) validation will flag error and visual editor will be disabled, according to design, until user fixes error in code
* Adding multiple inputs to a port will insert `MultipleInputFeatureRequirement`
* single scalar input -> scalar port works on CWLtool
* multiple inputs -> array port with linkmerge_flattened works on CWLtool

**Comment by ivanbatic** [comment]
> "None" option has been added to LinkMerge dropdown

What does "none" do? Per specification, you cannot have neither of the options, because 
> If not specified, the default method is "merge_nested"

Dropdown was implemented in a way that shows you "merge_nested" even if there's nothing in the document because that's how the execution will behave, but then the first time you explicitly select something from that dropdown, it will store the "linkMerge" property as whichever value you chose.

**Comment by kaushik-work** [comment]
Hi @ivanbatic ! None is for the case when we have single input to single port and we
- Don't need LinkMerge
- The inputs should not be sent as a list
